// node-pdfplumber/dist/chunk-22OGOIZS.js
import {
  findWordsInBBox
} from "./chunk-4FLXWJPM.js";

// src/utils/geometry.ts
function pointInBBox(point, bbox, tolerance = 0) {
  const [x, y] = point;
  return x >= bbox.x0 - tolerance && x <= bbox.x1 + tolerance && y >= bbox.y0 - tolerance && y <= bbox.y1 + tolerance;
}
function bboxIntersection(bbox1, bbox2) {
  const x0 = Math.max(bbox1.x0, bbox2.x0);
  const y0 = Math.max(bbox1.y0, bbox2.y0);
  const x1 = Math.min(bbox1.x1, bbox2.x1);
  const y1 = Math.min(bbox1.y1, bbox2.y1);
  if (x0 > x1 || y0 > y1) {
    return null;
  }
  return { x0, y0, x1, y1 };
}
function mergeBBoxes(bboxes) {
  if (bboxes.length === 0) return null;
  return {
    x0: Math.min(...bboxes.map((b) => b.x0)),
    y0: Math.min(...bboxes.map((b) => b.y0)),
    x1: Math.max(...bboxes.map((b) => b.x1)),
    y1: Math.max(...bboxes.map((b) => b.y1))
  };
}
function isHorizontalLine(x0, y0, x1, y1, tolerance = 3) {
  return Math.abs(y0 - y1) <= tolerance;
}
function isVerticalLine(x0, y0, x1, y1, tolerance = 3) {
  return Math.abs(x0 - x1) <= tolerance;
}
function transformPoint(point, matrix) {
  const [x, y] = point;
  const [a, b, c, d, e, f] = matrix;
  return [a * x + c * y + e, b * x + d * y + f];
}
function multiplyMatrices(m1, m2) {
  const [a1, b1, c1, d1, e1, f1] = m1;
  const [a2, b2, c2, d2, e2, f2] = m2;
  return [
    a1 * a2 + c1 * b2,
    b1 * a2 + d1 * b2,
    a1 * c2 + c1 * d2,
    b1 * c2 + d1 * d2,
    a1 * e2 + c1 * f2 + e1,
    b1 * e2 + d1 * f2 + f1
  ];
}
function identityMatrix() {
  return [1, 0, 0, 1, 0, 0];
}
function clusterNumbers(numbers, tolerance) {
  if (numbers.length === 0) return [];
  const sorted = [...numbers].sort((a, b) => a - b);
  const clusters = [[sorted[0]]];
  for (let i = 1; i < sorted.length; i++) {
    const current = sorted[i];
    const lastCluster = clusters[clusters.length - 1];
    const lastValue = lastCluster[lastCluster.length - 1];
    if (current - lastValue <= tolerance) {
      lastCluster.push(current);
    } else {
      clusters.push([current]);
    }
  }
  return clusters;
}
function clusterAverage(cluster) {
  return cluster.reduce((a, b) => a + b, 0) / cluster.length;
}

// src/table-finder.ts
var DEFAULT_SETTINGS = {
  verticalStrategy: "lines",
  horizontalStrategy: "lines",
  snapTolerance: 3,
  joinTolerance: 3,
  edgeMinLength: 3,
  minWords: 1,
  expandTableArea: 0
};
function extractTables(words, lines, rects, pageBox, settings) {
  const opts = { ...DEFAULT_SETTINGS, ...settings };
  if (opts.verticalStrategy === "lines" || opts.horizontalStrategy === "lines") {
    return extractTablesFromLines(words, lines, rects, pageBox, opts);
  } else {
    return extractTablesFromText(words, pageBox, opts);
  }
}
function extractTablesFromLines(words, lines, rects, pageBox, opts) {
  const allLines = [...lines, ...rectsToLines(rects)];
  const hLines = allLines.filter((line) => isHorizontalLine(line.x0, line.y0, line.x1, line.y1, 2));
  const vLines = allLines.filter((line) => isVerticalLine(line.x0, line.y0, line.x1, line.y1, 2));
  if (hLines.length < 2 || vLines.length < 2) {
    return [];
  }
  const hGrids = snapAndGroupLines(hLines, "h", opts.snapTolerance);
  const vGrids = snapAndGroupLines(vLines, "v", opts.snapTolerance);
  if (hGrids.length < 2 || vGrids.length < 2) {
    return [];
  }
  hGrids.sort((a, b) => a - b);
  vGrids.sort((a, b) => a - b);
  const tables = [];
  const tableRegions = detectTableRegions(hGrids, vGrids, words);
  for (const region of tableRegions) {
    const table = buildTable(
      words,
      hGrids.filter((y) => y >= region.y0 && y <= region.y1),
      vGrids.filter((x) => x >= region.x0 && x <= region.x1),
      opts
    );
    if (table) {
      tables.push(table);
    }
  }
  return tables;
}
function extractTablesFromText(words, pageBox, opts) {
  const rows = groupWordsIntoRows(words);
  if (rows.length < 2) {
    return [];
  }
  const allXPositions = [];
  for (const row of rows) {
    for (const word of row) {
      allXPositions.push(word.x0);
      allXPositions.push(word.x1);
    }
  }
  const xClusters = clusterNumbers(allXPositions, opts.snapTolerance || 3);
  const columnBoundaries = xClusters.map(clusterAverage).sort((a, b) => a - b);
  if (columnBoundaries.length < 2) {
    return [];
  }
  const cells = [];
  let cellIndex = 0;
  for (let rowIndex = 0; rowIndex < rows.length; rowIndex++) {
    const rowWords = rows[rowIndex];
    for (let colIndex = 0; colIndex < columnBoundaries.length - 1; colIndex++) {
      const colStart = columnBoundaries[colIndex];
      const colEnd = columnBoundaries[colIndex + 1];
      const cellWords = rowWords.filter(
        (w) => w.x0 >= colStart - opts.snapTolerance && w.x1 <= colEnd + opts.snapTolerance
      );
      const text = cellWords.map((w) => w.text).join(" ");
      const bbox = rowWords.length > 0 ? computeCellBBox(rowIndex, rows, colStart, colEnd) : null;
      if (bbox) {
        cells.push({
          text,
          bbox,
          rowIndex,
          colIndex
        });
      }
    }
  }
  if (cells.length === 0) {
    return [];
  }
  const table = {
    bbox: computeTableBBox(cells),
    cells,
    rows: rows.length,
    cols: columnBoundaries.length - 1
  };
  return [table];
}
function snapAndGroupLines(lines, direction, tolerance) {
  const coords = lines.map((line) => direction === "h" ? line.y0 : line.x0);
  const clusters = clusterNumbers(coords, tolerance);
  return clusters.map(clusterAverage);
}
function rectsToLines(rects) {
  const lines = [];
  for (const rect of rects) {
    if (rect.stroke) {
      const lineWidth = rect.lineWidth || 1;
      lines.push({
        x0: rect.x0,
        y0: rect.y0,
        x1: rect.x1,
        y1: rect.y0,
        width: lineWidth
      });
      lines.push({
        x0: rect.x0,
        y0: rect.y1,
        x1: rect.x1,
        y1: rect.y1,
        width: lineWidth
      });
      lines.push({
        x0: rect.x0,
        y0: rect.y0,
        x1: rect.x0,
        y1: rect.y1,
        width: lineWidth
      });
      lines.push({
        x0: rect.x1,
        y0: rect.y0,
        x1: rect.x1,
        y1: rect.y1,
        width: lineWidth
      });
    }
  }
  return lines;
}
function detectTableRegions(hGrids, vGrids, words) {
  if (hGrids.length < 2 || vGrids.length < 2) {
    return [];
  }
  return [
    {
      x0: vGrids[0],
      y0: hGrids[0],
      x1: vGrids[vGrids.length - 1],
      y1: hGrids[hGrids.length - 1]
    }
  ];
}
function buildTable(words, hGrids, vGrids, opts) {
  if (hGrids.length < 2 || vGrids.length < 2) {
    return null;
  }
  const rows = hGrids.length - 1;
  const cols = vGrids.length - 1;
  const cells = [];
  for (let rowIndex = 0; rowIndex < rows; rowIndex++) {
    const y0 = hGrids[rowIndex];
    const y1 = hGrids[rowIndex + 1];
    for (let colIndex = 0; colIndex < cols; colIndex++) {
      const x0 = vGrids[colIndex];
      const x1 = vGrids[colIndex + 1];
      const cellWords = findWordsInBBox(words, x0, y0, x1, y1);
      const text = cellWords.map((w) => w.text).join(" ");
      cells.push({
        text,
        bbox: { x0, y0, x1, y1 },
        rowIndex,
        colIndex
      });
    }
  }
  const bbox = {
    x0: vGrids[0],
    y0: hGrids[0],
    x1: vGrids[vGrids.length - 1],
    y1: hGrids[hGrids.length - 1]
  };
  return { bbox, cells, rows, cols };
}
function groupWordsIntoRows(words) {
  if (words.length === 0) return [];
  const sorted = [...words].sort((a, b) => a.y0 - b.y0);
  const rows = [[]];
  for (const word of sorted) {
    const lastRow = rows[rows.length - 1];
    if (lastRow.length === 0 || Math.abs(word.y0 - lastRow[0].y0) <= 5) {
      lastRow.push(word);
    } else {
      rows.push([word]);
    }
  }
  return rows.map((row) => row.sort((a, b) => a.x0 - b.x0));
}
function computeCellBBox(rowIndex, rows, x0, x1) {
  if (rowIndex >= rows.length) {
    return null;
  }
  const row = rows[rowIndex];
  if (row.length === 0) {
    return null;
  }
  const minY = Math.min(...row.map((w) => w.y0));
  const maxY = Math.max(...row.map((w) => w.y1));
  return { x0, y0: minY, x1, y1: maxY };
}
function computeTableBBox(cells) {
  if (cells.length === 0) {
    return { x0: 0, y0: 0, x1: 0, y1: 0 };
  }
  const bboxes = cells.map((c) => c.bbox);
  const merged = mergeBBoxes(bboxes);
  return merged || { x0: 0, y0: 0, x1: 0, y1: 0 };
}

export {
  pointInBBox,
  bboxIntersection,
  isHorizontalLine,
  isVerticalLine,
  transformPoint,
  multiplyMatrices,
  identityMatrix,
  extractTables
};
