// node-pdfplumber/dist/chunk-4FLXWJPM.js
// src/utils/text.ts
function textsToWords(items, pageHeight) {
  if (items.length === 0) return [];
  const words = [];
  for (const item of items) {
    if (!item.str || item.str.trim().length === 0) continue;
    const [a, b, c, d, e, f] = item.transform;
    const x0 = e;
    const y0_pdf = f;
    const x1 = e + (item.width || 0);
    const y1_pdf = f + (item.height || 0);
    const y1 = pageHeight - y0_pdf;
    const y0 = pageHeight - y1_pdf;
    const textParts = item.str.split(/(\s+)/);
    let currentX = x0;
    for (const part of textParts) {
      if (!part || part.length === 0) continue;
      if (part.trim().length === 0) {
        const spaceWidth = item.width / item.str.length * part.length;
        currentX += spaceWidth;
        continue;
      }
      const wordWidth = item.width / item.str.length * part.length;
      const wordX0 = currentX;
      const wordX1 = currentX + wordWidth;
      const word = {
        text: part,
        x0: wordX0,
        y0,
        x1: wordX1,
        y1,
        fontName: item.fontName
      };
      if (a !== 0) {
        word.fontSize = Math.abs(a);
      }
      words.push(word);
      currentX = wordX1;
    }
  }
  return words;
}
function sortWords(words) {
  return [...words].sort((a, b) => {
    if (Math.abs(a.y0 - b.y0) > 3) {
      return a.y0 - b.y0;
    }
    return a.x0 - b.x0;
  });
}
function groupWordsIntoLines(words, tolerance = 3) {
  if (words.length === 0) return [];
  const sorted = sortWords(words);
  const lines = [[]];
  for (const word of sorted) {
    const currentLine = lines[lines.length - 1];
    if (currentLine.length === 0) {
      currentLine.push(word);
    } else {
      const lastWord = currentLine[currentLine.length - 1];
      if (Math.abs(word.y0 - lastWord.y0) <= tolerance) {
        currentLine.push(word);
      } else {
        lines.push([word]);
      }
    }
  }
  return lines;
}
function wordsToText(words, layout = false) {
  if (words.length === 0) return "";
  if (!layout) {
    return words.map((w) => w.text).join(" ");
  }
  const lines = groupWordsIntoLines(words);
  const lineTexts = lines.map((lineWords) => {
    const sorted = [...lineWords].sort((a, b) => a.x0 - b.x0);
    let text = "";
    let lastX1 = sorted[0].x0;
    for (const word of sorted) {
      const gap = word.x0 - lastX1;
      if (gap > word.text.length * 2) {
        text += " ";
      }
      text += word.text;
      lastX1 = word.x1;
    }
    return text;
  });
  return lineTexts.join("\n");
}
function findWordsInBBox(words, x0, y0, x1, y1) {
  return words.filter(
    (w) => w.x0 >= x0 - 1 && w.x1 <= x1 + 1 && w.y0 >= y0 - 1 && w.y1 <= y1 + 1
  );
}
function findWordsMatching(words, pattern) {
  return words.filter((w) => pattern.test(w.text));
}

export {
  textsToWords,
  sortWords,
  groupWordsIntoLines,
  wordsToText,
  findWordsInBBox,
  findWordsMatching
};
