// node-pdfplumber/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/text.ts
var text_exports = {};
__export(text_exports, {
  findWordsInBBox: () => findWordsInBBox,
  findWordsMatching: () => findWordsMatching,
  groupWordsIntoLines: () => groupWordsIntoLines,
  sortWords: () => sortWords,
  textsToWords: () => textsToWords,
  wordsToText: () => wordsToText
});
function textsToWords(items, pageHeight) {
  if (items.length === 0) return [];
  const words = [];
  for (const item of items) {
    if (!item.str || item.str.trim().length === 0) continue;
    const [a, b, c, d, e, f] = item.transform;
    const x0 = e;
    const y0_pdf = f;
    const x1 = e + (item.width || 0);
    const y1_pdf = f + (item.height || 0);
    const y1 = pageHeight - y0_pdf;
    const y0 = pageHeight - y1_pdf;
    const textParts = item.str.split(/(\s+)/);
    let currentX = x0;
    for (const part of textParts) {
      if (!part || part.length === 0) continue;
      if (part.trim().length === 0) {
        const spaceWidth = item.width / item.str.length * part.length;
        currentX += spaceWidth;
        continue;
      }
      const wordWidth = item.width / item.str.length * part.length;
      const wordX0 = currentX;
      const wordX1 = currentX + wordWidth;
      const word = {
        text: part,
        x0: wordX0,
        y0,
        x1: wordX1,
        y1,
        fontName: item.fontName
      };
      if (a !== 0) {
        word.fontSize = Math.abs(a);
      }
      words.push(word);
      currentX = wordX1;
    }
  }
  return words;
}
function sortWords(words) {
  return [...words].sort((a, b) => {
    if (Math.abs(a.y0 - b.y0) > 3) {
      return a.y0 - b.y0;
    }
    return a.x0 - b.x0;
  });
}
function groupWordsIntoLines(words, tolerance = 3) {
  if (words.length === 0) return [];
  const sorted = sortWords(words);
  const lines = [[]];
  for (const word of sorted) {
    const currentLine = lines[lines.length - 1];
    if (currentLine.length === 0) {
      currentLine.push(word);
    } else {
      const lastWord = currentLine[currentLine.length - 1];
      if (Math.abs(word.y0 - lastWord.y0) <= tolerance) {
        currentLine.push(word);
      } else {
        lines.push([word]);
      }
    }
  }
  return lines;
}
function wordsToText(words, layout = false) {
  if (words.length === 0) return "";
  if (!layout) {
    return words.map((w) => w.text).join(" ");
  }
  const lines = groupWordsIntoLines(words);
  const lineTexts = lines.map((lineWords) => {
    const sorted = [...lineWords].sort((a, b) => a.x0 - b.x0);
    let text = "";
    let lastX1 = sorted[0].x0;
    for (const word of sorted) {
      const gap = word.x0 - lastX1;
      if (gap > word.text.length * 2) {
        text += " ";
      }
      text += word.text;
      lastX1 = word.x1;
    }
    return text;
  });
  return lineTexts.join("\n");
}
function findWordsInBBox(words, x0, y0, x1, y1) {
  return words.filter(
    (w) => w.x0 >= x0 - 1 && w.x1 <= x1 + 1 && w.y0 >= y0 - 1 && w.y1 <= y1 + 1
  );
}
function findWordsMatching(words, pattern) {
  return words.filter((w) => pattern.test(w.text));
}
var init_text = __esm({
  "src/utils/text.ts"() {
    "use strict";
  }
});

// src/utils/geometry.ts
function pointInBBox(point, bbox, tolerance = 0) {
  const [x, y] = point;
  return x >= bbox.x0 - tolerance && x <= bbox.x1 + tolerance && y >= bbox.y0 - tolerance && y <= bbox.y1 + tolerance;
}
function bboxIntersection(bbox1, bbox2) {
  const x0 = Math.max(bbox1.x0, bbox2.x0);
  const y0 = Math.max(bbox1.y0, bbox2.y0);
  const x1 = Math.min(bbox1.x1, bbox2.x1);
  const y1 = Math.min(bbox1.y1, bbox2.y1);
  if (x0 > x1 || y0 > y1) {
    return null;
  }
  return { x0, y0, x1, y1 };
}
function mergeBBoxes(bboxes) {
  if (bboxes.length === 0) return null;
  return {
    x0: Math.min(...bboxes.map((b) => b.x0)),
    y0: Math.min(...bboxes.map((b) => b.y0)),
    x1: Math.max(...bboxes.map((b) => b.x1)),
    y1: Math.max(...bboxes.map((b) => b.y1))
  };
}
function isHorizontalLine(x0, y0, x1, y1, tolerance = 3) {
  return Math.abs(y0 - y1) <= tolerance;
}
function isVerticalLine(x0, y0, x1, y1, tolerance = 3) {
  return Math.abs(x0 - x1) <= tolerance;
}
function transformPoint(point, matrix) {
  const [x, y] = point;
  const [a, b, c, d, e, f] = matrix;
  return [a * x + c * y + e, b * x + d * y + f];
}
function multiplyMatrices(m1, m2) {
  const [a1, b1, c1, d1, e1, f1] = m1;
  const [a2, b2, c2, d2, e2, f2] = m2;
  return [
    a1 * a2 + c1 * b2,
    b1 * a2 + d1 * b2,
    a1 * c2 + c1 * d2,
    b1 * c2 + d1 * d2,
    a1 * e2 + c1 * f2 + e1,
    b1 * e2 + d1 * f2 + f1
  ];
}
function identityMatrix() {
  return [1, 0, 0, 1, 0, 0];
}
function clusterNumbers(numbers, tolerance) {
  if (numbers.length === 0) return [];
  const sorted = [...numbers].sort((a, b) => a - b);
  const clusters = [[sorted[0]]];
  for (let i = 1; i < sorted.length; i++) {
    const current = sorted[i];
    const lastCluster = clusters[clusters.length - 1];
    const lastValue = lastCluster[lastCluster.length - 1];
    if (current - lastValue <= tolerance) {
      lastCluster.push(current);
    } else {
      clusters.push([current]);
    }
  }
  return clusters;
}
function clusterAverage(cluster) {
  return cluster.reduce((a, b) => a + b, 0) / cluster.length;
}
var init_geometry = __esm({
  "src/utils/geometry.ts"() {
    "use strict";
  }
});

// src/table-finder.ts
var table_finder_exports = {};
__export(table_finder_exports, {
  extractTables: () => extractTables
});
function extractTables(words, lines, rects, pageBox, settings) {
  const opts = { ...DEFAULT_SETTINGS, ...settings };
  if (opts.verticalStrategy === "lines" || opts.horizontalStrategy === "lines") {
    return extractTablesFromLines(words, lines, rects, pageBox, opts);
  } else {
    return extractTablesFromText(words, pageBox, opts);
  }
}
function extractTablesFromLines(words, lines, rects, pageBox, opts) {
  const allLines = [...lines, ...rectsToLines(rects)];
  const hLines = allLines.filter((line) => isHorizontalLine(line.x0, line.y0, line.x1, line.y1, 2));
  const vLines = allLines.filter((line) => isVerticalLine(line.x0, line.y0, line.x1, line.y1, 2));
  if (hLines.length < 2 || vLines.length < 2) {
    return [];
  }
  const hGrids = snapAndGroupLines(hLines, "h", opts.snapTolerance);
  const vGrids = snapAndGroupLines(vLines, "v", opts.snapTolerance);
  if (hGrids.length < 2 || vGrids.length < 2) {
    return [];
  }
  hGrids.sort((a, b) => a - b);
  vGrids.sort((a, b) => a - b);
  const tables = [];
  const tableRegions = detectTableRegions(hGrids, vGrids, words);
  for (const region of tableRegions) {
    const table = buildTable(
      words,
      hGrids.filter((y) => y >= region.y0 && y <= region.y1),
      vGrids.filter((x) => x >= region.x0 && x <= region.x1),
      opts
    );
    if (table) {
      tables.push(table);
    }
  }
  return tables;
}
function extractTablesFromText(words, pageBox, opts) {
  const rows = groupWordsIntoRows(words);
  if (rows.length < 2) {
    return [];
  }
  const allXPositions = [];
  for (const row of rows) {
    for (const word of row) {
      allXPositions.push(word.x0);
      allXPositions.push(word.x1);
    }
  }
  const xClusters = clusterNumbers(allXPositions, opts.snapTolerance || 3);
  const columnBoundaries = xClusters.map(clusterAverage).sort((a, b) => a - b);
  if (columnBoundaries.length < 2) {
    return [];
  }
  const cells = [];
  let cellIndex = 0;
  for (let rowIndex = 0; rowIndex < rows.length; rowIndex++) {
    const rowWords = rows[rowIndex];
    for (let colIndex = 0; colIndex < columnBoundaries.length - 1; colIndex++) {
      const colStart = columnBoundaries[colIndex];
      const colEnd = columnBoundaries[colIndex + 1];
      const cellWords = rowWords.filter(
        (w) => w.x0 >= colStart - opts.snapTolerance && w.x1 <= colEnd + opts.snapTolerance
      );
      const text = cellWords.map((w) => w.text).join(" ");
      const bbox = rowWords.length > 0 ? computeCellBBox(rowIndex, rows, colStart, colEnd) : null;
      if (bbox) {
        cells.push({
          text,
          bbox,
          rowIndex,
          colIndex
        });
      }
    }
  }
  if (cells.length === 0) {
    return [];
  }
  const table = {
    bbox: computeTableBBox(cells),
    cells,
    rows: rows.length,
    cols: columnBoundaries.length - 1
  };
  return [table];
}
function snapAndGroupLines(lines, direction, tolerance) {
  const coords = lines.map((line) => direction === "h" ? line.y0 : line.x0);
  const clusters = clusterNumbers(coords, tolerance);
  return clusters.map(clusterAverage);
}
function rectsToLines(rects) {
  const lines = [];
  for (const rect of rects) {
    if (rect.stroke) {
      const lineWidth = rect.lineWidth || 1;
      lines.push({
        x0: rect.x0,
        y0: rect.y0,
        x1: rect.x1,
        y1: rect.y0,
        width: lineWidth
      });
      lines.push({
        x0: rect.x0,
        y0: rect.y1,
        x1: rect.x1,
        y1: rect.y1,
        width: lineWidth
      });
      lines.push({
        x0: rect.x0,
        y0: rect.y0,
        x1: rect.x0,
        y1: rect.y1,
        width: lineWidth
      });
      lines.push({
        x0: rect.x1,
        y0: rect.y0,
        x1: rect.x1,
        y1: rect.y1,
        width: lineWidth
      });
    }
  }
  return lines;
}
function detectTableRegions(hGrids, vGrids, words) {
  if (hGrids.length < 2 || vGrids.length < 2) {
    return [];
  }
  return [
    {
      x0: vGrids[0],
      y0: hGrids[0],
      x1: vGrids[vGrids.length - 1],
      y1: hGrids[hGrids.length - 1]
    }
  ];
}
function buildTable(words, hGrids, vGrids, opts) {
  if (hGrids.length < 2 || vGrids.length < 2) {
    return null;
  }
  const rows = hGrids.length - 1;
  const cols = vGrids.length - 1;
  const cells = [];
  for (let rowIndex = 0; rowIndex < rows; rowIndex++) {
    const y0 = hGrids[rowIndex];
    const y1 = hGrids[rowIndex + 1];
    for (let colIndex = 0; colIndex < cols; colIndex++) {
      const x0 = vGrids[colIndex];
      const x1 = vGrids[colIndex + 1];
      const cellWords = findWordsInBBox(words, x0, y0, x1, y1);
      const text = cellWords.map((w) => w.text).join(" ");
      cells.push({
        text,
        bbox: { x0, y0, x1, y1 },
        rowIndex,
        colIndex
      });
    }
  }
  const bbox = {
    x0: vGrids[0],
    y0: hGrids[0],
    x1: vGrids[vGrids.length - 1],
    y1: hGrids[hGrids.length - 1]
  };
  return { bbox, cells, rows, cols };
}
function groupWordsIntoRows(words) {
  if (words.length === 0) return [];
  const sorted = [...words].sort((a, b) => a.y0 - b.y0);
  const rows = [[]];
  for (const word of sorted) {
    const lastRow = rows[rows.length - 1];
    if (lastRow.length === 0 || Math.abs(word.y0 - lastRow[0].y0) <= 5) {
      lastRow.push(word);
    } else {
      rows.push([word]);
    }
  }
  return rows.map((row) => row.sort((a, b) => a.x0 - b.x0));
}
function computeCellBBox(rowIndex, rows, x0, x1) {
  if (rowIndex >= rows.length) {
    return null;
  }
  const row = rows[rowIndex];
  if (row.length === 0) {
    return null;
  }
  const minY = Math.min(...row.map((w) => w.y0));
  const maxY = Math.max(...row.map((w) => w.y1));
  return { x0, y0: minY, x1, y1: maxY };
}
function computeTableBBox(cells) {
  if (cells.length === 0) {
    return { x0: 0, y0: 0, x1: 0, y1: 0 };
  }
  const bboxes = cells.map((c) => c.bbox);
  const merged = mergeBBoxes(bboxes);
  return merged || { x0: 0, y0: 0, x1: 0, y1: 0 };
}
var DEFAULT_SETTINGS;
var init_table_finder = __esm({
  "src/table-finder.ts"() {
    "use strict";
    init_geometry();
    init_text();
    DEFAULT_SETTINGS = {
      verticalStrategy: "lines",
      horizontalStrategy: "lines",
      snapTolerance: 3,
      joinTolerance: 3,
      edgeMinLength: 3,
      minWords: 1,
      expandTableArea: 0
    };
  }
});

// src/index.ts
var index_exports = {};
__export(index_exports, {
  CroppedPage: () => CroppedPage,
  PDFDocument: () => PDFDocument,
  PDFPage: () => PDFPage,
  PDFPlumber: () => PDFPlumber
});
module.exports = __toCommonJS(index_exports);

// src/pdfplumber.ts
var import_promises = require("fs/promises");
var import_pdf = require("pdfjs-dist/legacy/build/pdf.mjs");

// src/page.ts
init_text();

// src/utils/operator-list.ts
init_geometry();
var PDFJS_OPS = {
  save: 92,
  restore: 193,
  transform: 204,
  paintPath: 195,
  fill: 113,
  stroke: 201,
  fillStroke: 114,
  eoFill: 112,
  eoFillStroke: 115,
  closePath: 82,
  rectangle: 196,
  moveTo: 178,
  lineTo: 179,
  curveTo: 104,
  curveTo2: 105,
  curveTo3: 106,
  constructPath: 88,
  paintImageXObject: 182,
  paintInlineImageXObject: 183,
  setFillColor: 10,
  setStrokeColor: 200,
  setLineWidth: 175
};
var MatrixStack = class {
  constructor() {
    this.stack = [identityMatrix()];
  }
  current() {
    return this.stack[this.stack.length - 1];
  }
  save() {
    this.stack.push([...this.current()]);
  }
  restore() {
    if (this.stack.length > 1) {
      this.stack.pop();
    }
  }
  transform(a, b, c, d, e, f) {
    const current = this.current();
    const newMatrix = multiplyMatrices([a, b, c, d, e, f], current);
    this.stack[this.stack.length - 1] = newMatrix;
  }
};
var PathBuilder = class {
  constructor() {
    this.points = [];
    this.operators = [];
  }
  moveTo(x, y, matrix) {
    const p = transformPoint([x, y], matrix);
    this.points = [p];
    this.operators = ["moveTo"];
  }
  lineTo(x, y, matrix) {
    const p = transformPoint([x, y], matrix);
    if (this.points.length > 0) {
      this.points.push(p);
      this.operators.push("lineTo");
    }
  }
  curveTo(x1, y1, x2, y2, x3, y3, matrix) {
    const p = transformPoint([x3, y3], matrix);
    if (this.points.length > 0) {
      this.points.push(p);
      this.operators.push("curveTo");
    }
  }
  closePath() {
    if (this.points.length > 0) {
      this.operators.push("closePath");
    }
  }
  isEmpty() {
    return this.points.length === 0;
  }
  clear() {
    this.points = [];
    this.operators = [];
  }
};
async function extractGraphicsFromOperatorList(operatorList, pageHeight, pageWidth) {
  const lines = [];
  const rects = [];
  if (!operatorList || !operatorList.fnArray) {
    return { lines, rects };
  }
  const { fnArray, argsArray } = operatorList;
  const matrixStack = new MatrixStack();
  const pathBuilder = new PathBuilder();
  let currentStrokeWidth = 1;
  let isFilling = false;
  let isStroking = false;
  for (let i = 0; i < fnArray.length; i++) {
    const op = fnArray[i];
    const args = argsArray[i] || [];
    switch (op) {
      case PDFJS_OPS.save:
        matrixStack.save();
        break;
      case PDFJS_OPS.restore:
        matrixStack.restore();
        break;
      case PDFJS_OPS.transform: {
        const [a, b, c, d, e, f] = args;
        matrixStack.transform(a || 0, b || 0, c || 0, d || 0, e || 0, f || 0);
        break;
      }
      case PDFJS_OPS.setLineWidth:
        currentStrokeWidth = args[0] || 1;
        break;
      case PDFJS_OPS.moveTo: {
        const [x, y] = args;
        pathBuilder.moveTo(x, y, matrixStack.current());
        break;
      }
      case PDFJS_OPS.lineTo: {
        const [x, y] = args;
        pathBuilder.lineTo(x, y, matrixStack.current());
        break;
      }
      case PDFJS_OPS.curveTo: {
        const [x1, y1, x2, y2, x3, y3] = args;
        pathBuilder.curveTo(x1, y1, x2, y2, x3, y3, matrixStack.current());
        break;
      }
      case PDFJS_OPS.closePath:
        pathBuilder.closePath();
        break;
      case PDFJS_OPS.rectangle: {
        const [x, y, width, height] = args;
        const matrix = matrixStack.current();
        const p1 = transformPoint([x, y], matrix);
        const p2 = transformPoint([x + width, y + height], matrix);
        const rect = {
          x0: Math.min(p1[0], p2[0]),
          y0: Math.min(p1[1], p2[1]),
          x1: Math.max(p1[0], p2[0]),
          y1: Math.max(p1[1], p2[1]),
          stroke: true,
          lineWidth: currentStrokeWidth
        };
        const flippedRect = {
          x0: rect.x0,
          y0: pageHeight - rect.y1,
          x1: rect.x1,
          y1: pageHeight - rect.y0,
          stroke: rect.stroke,
          lineWidth: rect.lineWidth
        };
        rects.push(flippedRect);
        pathBuilder.clear();
        break;
      }
      case PDFJS_OPS.fill:
      case PDFJS_OPS.fillStroke:
      case PDFJS_OPS.eoFill:
      case PDFJS_OPS.eoFillStroke:
        isFilling = true;
        isStroking = op === PDFJS_OPS.fillStroke || op === PDFJS_OPS.eoFillStroke;
        processPath(pathBuilder, lines, rects, pageHeight, currentStrokeWidth, isStroking, isFilling);
        pathBuilder.clear();
        isFilling = false;
        break;
      case PDFJS_OPS.stroke:
        isStroking = true;
        processPath(pathBuilder, lines, rects, pageHeight, currentStrokeWidth, isStroking, false);
        pathBuilder.clear();
        isStroking = false;
        break;
      case PDFJS_OPS.constructPath: {
        break;
      }
    }
  }
  const uniqueLines = deduplicateLines(lines, 2);
  return { lines: uniqueLines, rects };
}
function processPath(pathBuilder, lines, rects, pageHeight, strokeWidth, isStroking, isFilling) {
  if (pathBuilder.isEmpty()) return;
  const points = pathBuilder.points;
  if ((points.length === 4 || points.length === 5) && pathBuilder.operators.includes("closePath")) {
    const rect = detectRectangle(points);
    if (rect) {
      const flipped = {
        x0: rect.x0,
        y0: pageHeight - rect.y1,
        x1: rect.x1,
        y1: pageHeight - rect.y0,
        stroke: isStroking,
        fill: isFilling,
        lineWidth: strokeWidth
      };
      rects.push(flipped);
      return;
    }
  }
  if (isStroking && points.length >= 2) {
    for (let i = 0; i < points.length - 1; i++) {
      const [x0, y0] = points[i];
      const [x1, y1] = points[i + 1];
      const line = {
        x0,
        y0: pageHeight - y1,
        x1,
        y1: pageHeight - y0,
        width: strokeWidth
      };
      if (isHorizontalLine(x0, y0, x1, y1, 1) || isVerticalLine(x0, y0, x1, y1, 1)) {
        lines.push(line);
      }
    }
  }
}
function detectRectangle(points) {
  if (points.length < 4) return null;
  const xs = points.map((p) => p[0]);
  const ys = points.map((p) => p[1]);
  const x0 = Math.min(...xs);
  const x1 = Math.max(...xs);
  const y0 = Math.min(...ys);
  const y1 = Math.max(...ys);
  const tolerance = 2;
  const corners = [
    [x0, y0],
    [x1, y0],
    [x1, y1],
    [x0, y1]
  ];
  const matches = points.filter(
    (p) => corners.some((c) => Math.abs(p[0] - c[0]) < tolerance && Math.abs(p[1] - c[1]) < tolerance)
  );
  if (matches.length >= 3) {
    return { x0, y0, x1, y1 };
  }
  return null;
}
function deduplicateLines(lines, tolerance) {
  const unique = [];
  for (const line of lines) {
    const isDuplicate = unique.some(
      (existing) => Math.abs(existing.x0 - line.x0) < tolerance && Math.abs(existing.y0 - line.y0) < tolerance && Math.abs(existing.x1 - line.x1) < tolerance && Math.abs(existing.y1 - line.y1) < tolerance
    );
    if (!isDuplicate) {
      unique.push(line);
    }
  }
  return unique;
}

// src/page.ts
init_table_finder();

// src/cropped-page.ts
init_geometry();
var CroppedPage = class _CroppedPage {
  constructor(page, cropbox) {
    this.page = page;
    this.cropbox = cropbox;
  }
  get number() {
    return this.page.number;
  }
  get width() {
    return this.page.width;
  }
  get height() {
    return this.page.height;
  }
  /**
   * Get crop bounding box
   */
  get bbox() {
    return this.cropbox;
  }
  /**
   * Extract words to filter by crop region
   */
  async extractWords() {
    const words = await this.page.extractWords();
    return words.filter((word) => this.pointInCrop(word.x0, word.y0));
  }
  /**
   * Get chars to filter by crop region
   */
  async getChars() {
    const chars = await this.page.getChars();
    return chars.filter((char) => this.pointInCrop(char.x0, char.y0));
  }
  /**
   * Get rects to filter by crop region
   */
  async getRects() {
    const rects = await this.page.getRects();
    return rects.map((rect) => {
      const intersection = bboxIntersection(rect, this.cropbox);
      return intersection || rect;
    }).filter((rect) => {
      return rect.x0 >= this.cropbox.x0 - 1 && rect.x1 <= this.cropbox.x1 + 1 && rect.y0 >= this.cropbox.y0 - 1 && rect.y1 <= this.cropbox.y1 + 1;
    });
  }
  /**
   * Get lines to filter by crop region
   */
  async getLines() {
    const lines = await this.page.getLines();
    return lines.filter((line) => {
      const p1InCrop = this.pointInCrop(line.x0, line.y0);
      const p2InCrop = this.pointInCrop(line.x1, line.y1);
      return p1InCrop || p2InCrop;
    });
  }
  /**
   * Extract text from cropped region
   */
  async extractText(layout = false) {
    const words = await this.extractWords();
    const { wordsToText: wordsToText2 } = await Promise.resolve().then(() => (init_text(), text_exports));
    return wordsToText2(words, layout);
  }
  /**
   * Extract table from cropped region
   */
  async extractTable(settings) {
    const tables = await this.extractTables(settings);
    return tables.length > 0 ? tables[0] : null;
  }
  /**
   * Extract tables from cropped data
   */
  async extractTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    const { extractTables: extract } = await Promise.resolve().then(() => (init_table_finder(), table_finder_exports));
    const tables = extract(words, lines, rects, this.cropbox, settings);
    return tables.map((table) => {
      const result = Array.from(
        { length: table.rows },
        () => Array.from({ length: table.cols }, () => "")
      );
      for (let i = 0; i < table.cells.length; i++) {
        const cell = table.cells[i];
        if (cell && result[cell.rowIndex]) {
          result[cell.rowIndex][cell.colIndex] = cell.text;
        }
      }
      return result;
    });
  }
  /**
   * Find tables from cropped data
   */
  async findTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    const { extractTables: extract } = await Promise.resolve().then(() => (init_table_finder(), table_finder_exports));
    return extract(words, lines, rects, this.cropbox, settings);
  }
  /**
   * Check if point is within crop box
   */
  pointInCrop(x, y) {
    return pointInBBox([x, y], this.cropbox, 1);
  }
  /**
   * Crop to create new CroppedPage with intersection
   */
  crop(bbox) {
    const intersection = bboxIntersection(this.cropbox, bbox);
    return new _CroppedPage(this.page, intersection || bbox);
  }
};

// src/page.ts
var PDFPage = class {
  constructor(document, pageProxy, pageNumber) {
    this._width = 0;
    this._height = 0;
    // Caches
    this._textContent = null;
    this._words = null;
    this._lines = null;
    this._rects = null;
    this.document = document;
    this.pageProxy = pageProxy;
    this.pageNumber = pageNumber;
    if (pageProxy.view) {
      const [, , width, height] = pageProxy.view;
      this._width = width;
      this._height = height;
    }
  }
  /**
   * Get page number (1-indexed)
   */
  get number() {
    return this.pageNumber;
  }
  /**
   * Get page width in user space units
   */
  get width() {
    return this._width;
  }
  /**
   * Get page height in user space units
   */
  get height() {
    return this._height;
  }
  /**
   * Get full page bounding box
   */
  get bbox() {
    return {
      x0: 0,
      y0: 0,
      x1: this._width,
      y1: this._height
    };
  }
  /**
   * Get raw text content from pdfjs
   */
  async getTextContent() {
    if (this._textContent !== null) {
      return this._textContent;
    }
    try {
      const content = await this.pageProxy.getTextContent();
      this._textContent = content.items || [];
      return this._textContent;
    } catch (e) {
      console.error("Error extracting text content:", e);
      this._textContent = [];
      return [];
    }
  }
  /**
   * Extract words with bounding boxes
   */
  async extractWords() {
    if (this._words !== null) {
      return this._words;
    }
    const textContent = await this.getTextContent();
    const words = textsToWords(textContent, this._height);
    this._words = sortWords(words);
    return this._words;
  }
  /**
   * Extract all text from page
   */
  async extractText(layout = false) {
    const words = await this.extractWords();
    return wordsToText(words, layout);
  }
  /**
   * Get character-level data (currently returns words split into chars)
   * For true character-level data, would need mupdf
   */
  async getChars() {
    const words = await this.extractWords();
    const chars = [];
    for (const word of words) {
      const charWidth = (word.x1 - word.x0) / word.text.length;
      for (let i = 0; i < word.text.length; i++) {
        chars.push({
          text: word.text[i],
          x0: word.x0 + i * charWidth,
          y0: word.y0,
          x1: word.x0 + (i + 1) * charWidth,
          y1: word.y1,
          fontName: word.fontName,
          fontSize: word.fontSize,
          isWhitespace: /\s/.test(word.text[i])
        });
      }
    }
    return chars;
  }
  /**
   * Get all rectangles on page
   */
  async getRects() {
    if (this._rects !== null) {
      return this._rects;
    }
    const opList = await this.pageProxy.getOperatorList();
    const { rects } = await extractGraphicsFromOperatorList(opList, this._height, this._width);
    this._rects = rects;
    return rects;
  }
  /**
   * Get all lines on page
   */
  async getLines() {
    if (this._lines !== null) {
      return this._lines;
    }
    const opList = await this.pageProxy.getOperatorList();
    const { lines } = await extractGraphicsFromOperatorList(opList, this._height, this._width);
    this._lines = lines;
    return lines;
  }
  /**
   * Extract the first/main table from page
   */
  async extractTable(settings) {
    const tables = await this.extractTables(settings);
    return tables.length > 0 ? tables[0] : null;
  }
  /**
   * Extract all tables from page
   */
  async extractTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    const tables = extractTables(words, lines, rects, this.bbox, settings);
    return tables.map((table) => {
      const result = Array.from(
        { length: table.rows },
        () => Array.from({ length: table.cols }, () => "")
      );
      for (let i = 0; i < table.cells.length; i++) {
        const cell = table.cells[i];
        if (cell && result[cell.rowIndex]) {
          result[cell.rowIndex][cell.colIndex] = cell.text;
        }
      }
      return result;
    });
  }
  /**
   * Find and return Table objects with metadata
   */
  async findTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    return extractTables(words, lines, rects, this.bbox, settings);
  }
  /**
   * Crop page to region and return CroppedPage
   */
  crop(bbox) {
    return new CroppedPage(this, bbox);
  }
  /**
   * Alias for crop()
   */
  within_bbox(bbox) {
    return this.crop(bbox);
  }
  /**
   * Get the underlying pdfjs page proxy (for advanced use)
   */
  getProxy() {
    return this.pageProxy;
  }
};

// src/document.ts
var PDFDocument = class _PDFDocument {
  constructor(pdfDocProxy) {
    // pdfjs-dist PDFDocumentProxy
    this._pages = [];
    this._metadata = {};
    this._pageCount = 0;
    this.pdfDocProxy = pdfDocProxy;
    this._pageCount = pdfDocProxy.numPages || 0;
  }
  /**
   * Factory method to create PDFDocument from pdfjs-dist PDFDocumentProxy
   */
  static async from(pdfDocProxy) {
    const doc = new _PDFDocument(pdfDocProxy);
    try {
      const metadata = await pdfDocProxy.getMetadata();
      if (metadata) {
        doc._metadata = parseMetadata(metadata);
      }
    } catch (e) {
    }
    for (let i = 1; i <= doc._pageCount; i++) {
      const pageProxy = await pdfDocProxy.getPage(i);
      const page = new PDFPage(doc, pageProxy, i);
      doc._pages.push(page);
    }
    return doc;
  }
  /**
   * Get all pages
   */
  get pages() {
    return this._pages;
  }
  /**
   * Get number of pages
   */
  get pageCount() {
    return this._pageCount;
  }
  /**
   * Get document metadata
   */
  get metadata() {
    return this._metadata;
  }
  /**
   * Get internal pdfjs proxy (for advanced use)
   */
  getProxy() {
    return this.pdfDocProxy;
  }
  /**
   * Close the document and clean up resources
   */
  close() {
    if (this.pdfDocProxy && this.pdfDocProxy.destroy) {
      this.pdfDocProxy.destroy();
    }
  }
};
function parseMetadata(metadata) {
  const result = {};
  const info = metadata.info || {};
  if (info.Title) result.title = info.Title;
  if (info.Author) result.author = info.Author;
  if (info.Subject) result.subject = info.Subject;
  if (info.Keywords) result.keywords = info.Keywords;
  if (info.Creator) result.creator = info.Creator;
  if (info.Producer) result.producer = info.Producer;
  if (info.CreationDate) {
    result.creationDate = parseDate(info.CreationDate);
  }
  if (info.ModDate) {
    result.modificationDate = parseDate(info.ModDate);
  }
  for (const [key, value] of Object.entries(info)) {
    if (!["Title", "Author", "Subject", "Keywords", "Creator", "Producer", "CreationDate", "ModDate"].includes(
      key
    )) {
      result[key] = value;
    }
  }
  return result;
}
function parseDate(dateStr) {
  if (dateStr instanceof Date) {
    return dateStr;
  }
  if (typeof dateStr !== "string") {
    return void 0;
  }
  const cleanStr = dateStr.replace(/^D:/, "");
  const year = parseInt(cleanStr.substring(0, 4), 10);
  const month = parseInt(cleanStr.substring(4, 6), 10) || 1;
  const day = parseInt(cleanStr.substring(6, 8), 10) || 1;
  const hour = parseInt(cleanStr.substring(8, 10), 10) || 0;
  const minute = parseInt(cleanStr.substring(10, 12), 10) || 0;
  const second = parseInt(cleanStr.substring(12, 14), 10) || 0;
  if (isNaN(year)) {
    return void 0;
  }
  return new Date(year, month - 1, day, hour, minute, second);
}

// src/pdfplumber.ts
var PDFPlumber = class _PDFPlumber {
  /**
   * Open a PDF from file path or buffer
   */
  static async open(input) {
    let data;
    if (typeof input === "string") {
      data = await (0, import_promises.readFile)(input);
    } else if (input instanceof Buffer) {
      data = new Uint8Array(input);
    } else {
      data = input;
    }
    return _PDFPlumber.fromBuffer(data);
  }
  /**
   * Load PDF from buffer or Uint8Array
   */
  static async fromBuffer(buffer) {
    const uint8array = buffer instanceof Buffer ? new Uint8Array(buffer) : buffer;
    const loadingTask = (0, import_pdf.getDocument)({
      data: uint8array,
      useSystemFonts: true,
      standardFontDataUrl: "https://cdn.jsdelivr.net/npm/pdfjs-dist@4.10.38/standard_fonts/"
    });
    try {
      const pdfDocProxy = await loadingTask.promise;
      return PDFDocument.from(pdfDocProxy);
    } catch (error) {
      loadingTask.destroy();
      throw error;
    }
  }
  /**
   * Load PDF from URL
   */
  static async fromURL(url) {
    const response = await fetch(url);
    const buffer = await response.arrayBuffer();
    return _PDFPlumber.fromBuffer(new Uint8Array(buffer));
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CroppedPage,
  PDFDocument,
  PDFPage,
  PDFPlumber
});
