// node-pdfplumber/dist/index.js
import {
  bboxIntersection,
  extractTables,
  identityMatrix,
  isHorizontalLine,
  isVerticalLine,
  multiplyMatrices,
  pointInBBox,
  transformPoint
} from "./chunk-22OGOIZS.js";
import {
  sortWords,
  textsToWords,
  wordsToText
} from "./chunk-4FLXWJPM.js";

// src/pdfplumber.ts
import { readFile } from "fs/promises";
import { getDocument } from "pdfjs-dist/legacy/build/pdf.mjs";

// src/utils/operator-list.ts
var PDFJS_OPS = {
  save: 92,
  restore: 193,
  transform: 204,
  paintPath: 195,
  fill: 113,
  stroke: 201,
  fillStroke: 114,
  eoFill: 112,
  eoFillStroke: 115,
  closePath: 82,
  rectangle: 196,
  moveTo: 178,
  lineTo: 179,
  curveTo: 104,
  curveTo2: 105,
  curveTo3: 106,
  constructPath: 88,
  paintImageXObject: 182,
  paintInlineImageXObject: 183,
  setFillColor: 10,
  setStrokeColor: 200,
  setLineWidth: 175
};
var MatrixStack = class {
  constructor() {
    this.stack = [identityMatrix()];
  }
  current() {
    return this.stack[this.stack.length - 1];
  }
  save() {
    this.stack.push([...this.current()]);
  }
  restore() {
    if (this.stack.length > 1) {
      this.stack.pop();
    }
  }
  transform(a, b, c, d, e, f) {
    const current = this.current();
    const newMatrix = multiplyMatrices([a, b, c, d, e, f], current);
    this.stack[this.stack.length - 1] = newMatrix;
  }
};
var PathBuilder = class {
  constructor() {
    this.points = [];
    this.operators = [];
  }
  moveTo(x, y, matrix) {
    const p = transformPoint([x, y], matrix);
    this.points = [p];
    this.operators = ["moveTo"];
  }
  lineTo(x, y, matrix) {
    const p = transformPoint([x, y], matrix);
    if (this.points.length > 0) {
      this.points.push(p);
      this.operators.push("lineTo");
    }
  }
  curveTo(x1, y1, x2, y2, x3, y3, matrix) {
    const p = transformPoint([x3, y3], matrix);
    if (this.points.length > 0) {
      this.points.push(p);
      this.operators.push("curveTo");
    }
  }
  closePath() {
    if (this.points.length > 0) {
      this.operators.push("closePath");
    }
  }
  isEmpty() {
    return this.points.length === 0;
  }
  clear() {
    this.points = [];
    this.operators = [];
  }
};
async function extractGraphicsFromOperatorList(operatorList, pageHeight, pageWidth) {
  const lines = [];
  const rects = [];
  if (!operatorList || !operatorList.fnArray) {
    return { lines, rects };
  }
  const { fnArray, argsArray } = operatorList;
  const matrixStack = new MatrixStack();
  const pathBuilder = new PathBuilder();
  let currentStrokeWidth = 1;
  let isFilling = false;
  let isStroking = false;
  for (let i = 0; i < fnArray.length; i++) {
    const op = fnArray[i];
    const args = argsArray[i] || [];
    switch (op) {
      case PDFJS_OPS.save:
        matrixStack.save();
        break;
      case PDFJS_OPS.restore:
        matrixStack.restore();
        break;
      case PDFJS_OPS.transform: {
        const [a, b, c, d, e, f] = args;
        matrixStack.transform(a || 0, b || 0, c || 0, d || 0, e || 0, f || 0);
        break;
      }
      case PDFJS_OPS.setLineWidth:
        currentStrokeWidth = args[0] || 1;
        break;
      case PDFJS_OPS.moveTo: {
        const [x, y] = args;
        pathBuilder.moveTo(x, y, matrixStack.current());
        break;
      }
      case PDFJS_OPS.lineTo: {
        const [x, y] = args;
        pathBuilder.lineTo(x, y, matrixStack.current());
        break;
      }
      case PDFJS_OPS.curveTo: {
        const [x1, y1, x2, y2, x3, y3] = args;
        pathBuilder.curveTo(x1, y1, x2, y2, x3, y3, matrixStack.current());
        break;
      }
      case PDFJS_OPS.closePath:
        pathBuilder.closePath();
        break;
      case PDFJS_OPS.rectangle: {
        const [x, y, width, height] = args;
        const matrix = matrixStack.current();
        const p1 = transformPoint([x, y], matrix);
        const p2 = transformPoint([x + width, y + height], matrix);
        const rect = {
          x0: Math.min(p1[0], p2[0]),
          y0: Math.min(p1[1], p2[1]),
          x1: Math.max(p1[0], p2[0]),
          y1: Math.max(p1[1], p2[1]),
          stroke: true,
          lineWidth: currentStrokeWidth
        };
        const flippedRect = {
          x0: rect.x0,
          y0: pageHeight - rect.y1,
          x1: rect.x1,
          y1: pageHeight - rect.y0,
          stroke: rect.stroke,
          lineWidth: rect.lineWidth
        };
        rects.push(flippedRect);
        pathBuilder.clear();
        break;
      }
      case PDFJS_OPS.fill:
      case PDFJS_OPS.fillStroke:
      case PDFJS_OPS.eoFill:
      case PDFJS_OPS.eoFillStroke:
        isFilling = true;
        isStroking = op === PDFJS_OPS.fillStroke || op === PDFJS_OPS.eoFillStroke;
        processPath(pathBuilder, lines, rects, pageHeight, currentStrokeWidth, isStroking, isFilling);
        pathBuilder.clear();
        isFilling = false;
        break;
      case PDFJS_OPS.stroke:
        isStroking = true;
        processPath(pathBuilder, lines, rects, pageHeight, currentStrokeWidth, isStroking, false);
        pathBuilder.clear();
        isStroking = false;
        break;
      case PDFJS_OPS.constructPath: {
        break;
      }
    }
  }
  const uniqueLines = deduplicateLines(lines, 2);
  return { lines: uniqueLines, rects };
}
function processPath(pathBuilder, lines, rects, pageHeight, strokeWidth, isStroking, isFilling) {
  if (pathBuilder.isEmpty()) return;
  const points = pathBuilder.points;
  if ((points.length === 4 || points.length === 5) && pathBuilder.operators.includes("closePath")) {
    const rect = detectRectangle(points);
    if (rect) {
      const flipped = {
        x0: rect.x0,
        y0: pageHeight - rect.y1,
        x1: rect.x1,
        y1: pageHeight - rect.y0,
        stroke: isStroking,
        fill: isFilling,
        lineWidth: strokeWidth
      };
      rects.push(flipped);
      return;
    }
  }
  if (isStroking && points.length >= 2) {
    for (let i = 0; i < points.length - 1; i++) {
      const [x0, y0] = points[i];
      const [x1, y1] = points[i + 1];
      const line = {
        x0,
        y0: pageHeight - y1,
        x1,
        y1: pageHeight - y0,
        width: strokeWidth
      };
      if (isHorizontalLine(x0, y0, x1, y1, 1) || isVerticalLine(x0, y0, x1, y1, 1)) {
        lines.push(line);
      }
    }
  }
}
function detectRectangle(points) {
  if (points.length < 4) return null;
  const xs = points.map((p) => p[0]);
  const ys = points.map((p) => p[1]);
  const x0 = Math.min(...xs);
  const x1 = Math.max(...xs);
  const y0 = Math.min(...ys);
  const y1 = Math.max(...ys);
  const tolerance = 2;
  const corners = [
    [x0, y0],
    [x1, y0],
    [x1, y1],
    [x0, y1]
  ];
  const matches = points.filter(
    (p) => corners.some((c) => Math.abs(p[0] - c[0]) < tolerance && Math.abs(p[1] - c[1]) < tolerance)
  );
  if (matches.length >= 3) {
    return { x0, y0, x1, y1 };
  }
  return null;
}
function deduplicateLines(lines, tolerance) {
  const unique = [];
  for (const line of lines) {
    const isDuplicate = unique.some(
      (existing) => Math.abs(existing.x0 - line.x0) < tolerance && Math.abs(existing.y0 - line.y0) < tolerance && Math.abs(existing.x1 - line.x1) < tolerance && Math.abs(existing.y1 - line.y1) < tolerance
    );
    if (!isDuplicate) {
      unique.push(line);
    }
  }
  return unique;
}

// src/cropped-page.ts
var CroppedPage = class _CroppedPage {
  constructor(page, cropbox) {
    this.page = page;
    this.cropbox = cropbox;
  }
  get number() {
    return this.page.number;
  }
  get width() {
    return this.page.width;
  }
  get height() {
    return this.page.height;
  }
  /**
   * Get crop bounding box
   */
  get bbox() {
    return this.cropbox;
  }
  /**
   * Extract words to filter by crop region
   */
  async extractWords() {
    const words = await this.page.extractWords();
    return words.filter((word) => this.pointInCrop(word.x0, word.y0));
  }
  /**
   * Get chars to filter by crop region
   */
  async getChars() {
    const chars = await this.page.getChars();
    return chars.filter((char) => this.pointInCrop(char.x0, char.y0));
  }
  /**
   * Get rects to filter by crop region
   */
  async getRects() {
    const rects = await this.page.getRects();
    return rects.map((rect) => {
      const intersection = bboxIntersection(rect, this.cropbox);
      return intersection || rect;
    }).filter((rect) => {
      return rect.x0 >= this.cropbox.x0 - 1 && rect.x1 <= this.cropbox.x1 + 1 && rect.y0 >= this.cropbox.y0 - 1 && rect.y1 <= this.cropbox.y1 + 1;
    });
  }
  /**
   * Get lines to filter by crop region
   */
  async getLines() {
    const lines = await this.page.getLines();
    return lines.filter((line) => {
      const p1InCrop = this.pointInCrop(line.x0, line.y0);
      const p2InCrop = this.pointInCrop(line.x1, line.y1);
      return p1InCrop || p2InCrop;
    });
  }
  /**
   * Extract text from cropped region
   */
  async extractText(layout = false) {
    const words = await this.extractWords();
    const { wordsToText: wordsToText2 } = await import("./text-OEGY66AY.js");
    return wordsToText2(words, layout);
  }
  /**
   * Extract table from cropped region
   */
  async extractTable(settings) {
    const tables = await this.extractTables(settings);
    return tables.length > 0 ? tables[0] : null;
  }
  /**
   * Extract tables from cropped data
   */
  async extractTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    const { extractTables: extract } = await import("./table-finder-ZBZGEXID.js");
    const tables = extract(words, lines, rects, this.cropbox, settings);
    return tables.map((table) => {
      const result = Array.from(
        { length: table.rows },
        () => Array.from({ length: table.cols }, () => "")
      );
      for (let i = 0; i < table.cells.length; i++) {
        const cell = table.cells[i];
        if (cell && result[cell.rowIndex]) {
          result[cell.rowIndex][cell.colIndex] = cell.text;
        }
      }
      return result;
    });
  }
  /**
   * Find tables from cropped data
   */
  async findTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    const { extractTables: extract } = await import("./table-finder-ZBZGEXID.js");
    return extract(words, lines, rects, this.cropbox, settings);
  }
  /**
   * Check if point is within crop box
   */
  pointInCrop(x, y) {
    return pointInBBox([x, y], this.cropbox, 1);
  }
  /**
   * Crop to create new CroppedPage with intersection
   */
  crop(bbox) {
    const intersection = bboxIntersection(this.cropbox, bbox);
    return new _CroppedPage(this.page, intersection || bbox);
  }
};

// src/page.ts
var PDFPage = class {
  constructor(document, pageProxy, pageNumber) {
    this._width = 0;
    this._height = 0;
    // Caches
    this._textContent = null;
    this._words = null;
    this._lines = null;
    this._rects = null;
    this.document = document;
    this.pageProxy = pageProxy;
    this.pageNumber = pageNumber;
    if (pageProxy.view) {
      const [, , width, height] = pageProxy.view;
      this._width = width;
      this._height = height;
    }
  }
  /**
   * Get page number (1-indexed)
   */
  get number() {
    return this.pageNumber;
  }
  /**
   * Get page width in user space units
   */
  get width() {
    return this._width;
  }
  /**
   * Get page height in user space units
   */
  get height() {
    return this._height;
  }
  /**
   * Get full page bounding box
   */
  get bbox() {
    return {
      x0: 0,
      y0: 0,
      x1: this._width,
      y1: this._height
    };
  }
  /**
   * Get raw text content from pdfjs
   */
  async getTextContent() {
    if (this._textContent !== null) {
      return this._textContent;
    }
    try {
      const content = await this.pageProxy.getTextContent();
      this._textContent = content.items || [];
      return this._textContent;
    } catch (e) {
      console.error("Error extracting text content:", e);
      this._textContent = [];
      return [];
    }
  }
  /**
   * Extract words with bounding boxes
   */
  async extractWords() {
    if (this._words !== null) {
      return this._words;
    }
    const textContent = await this.getTextContent();
    const words = textsToWords(textContent, this._height);
    this._words = sortWords(words);
    return this._words;
  }
  /**
   * Extract all text from page
   */
  async extractText(layout = false) {
    const words = await this.extractWords();
    return wordsToText(words, layout);
  }
  /**
   * Get character-level data (currently returns words split into chars)
   * For true character-level data, would need mupdf
   */
  async getChars() {
    const words = await this.extractWords();
    const chars = [];
    for (const word of words) {
      const charWidth = (word.x1 - word.x0) / word.text.length;
      for (let i = 0; i < word.text.length; i++) {
        chars.push({
          text: word.text[i],
          x0: word.x0 + i * charWidth,
          y0: word.y0,
          x1: word.x0 + (i + 1) * charWidth,
          y1: word.y1,
          fontName: word.fontName,
          fontSize: word.fontSize,
          isWhitespace: /\s/.test(word.text[i])
        });
      }
    }
    return chars;
  }
  /**
   * Get all rectangles on page
   */
  async getRects() {
    if (this._rects !== null) {
      return this._rects;
    }
    const opList = await this.pageProxy.getOperatorList();
    const { rects } = await extractGraphicsFromOperatorList(opList, this._height, this._width);
    this._rects = rects;
    return rects;
  }
  /**
   * Get all lines on page
   */
  async getLines() {
    if (this._lines !== null) {
      return this._lines;
    }
    const opList = await this.pageProxy.getOperatorList();
    const { lines } = await extractGraphicsFromOperatorList(opList, this._height, this._width);
    this._lines = lines;
    return lines;
  }
  /**
   * Extract the first/main table from page
   */
  async extractTable(settings) {
    const tables = await this.extractTables(settings);
    return tables.length > 0 ? tables[0] : null;
  }
  /**
   * Extract all tables from page
   */
  async extractTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    const tables = extractTables(words, lines, rects, this.bbox, settings);
    return tables.map((table) => {
      const result = Array.from(
        { length: table.rows },
        () => Array.from({ length: table.cols }, () => "")
      );
      for (let i = 0; i < table.cells.length; i++) {
        const cell = table.cells[i];
        if (cell && result[cell.rowIndex]) {
          result[cell.rowIndex][cell.colIndex] = cell.text;
        }
      }
      return result;
    });
  }
  /**
   * Find and return Table objects with metadata
   */
  async findTables(settings) {
    const words = await this.extractWords();
    const lines = await this.getLines();
    const rects = await this.getRects();
    return extractTables(words, lines, rects, this.bbox, settings);
  }
  /**
   * Crop page to region and return CroppedPage
   */
  crop(bbox) {
    return new CroppedPage(this, bbox);
  }
  /**
   * Alias for crop()
   */
  within_bbox(bbox) {
    return this.crop(bbox);
  }
  /**
   * Get the underlying pdfjs page proxy (for advanced use)
   */
  getProxy() {
    return this.pageProxy;
  }
};

// src/document.ts
var PDFDocument = class _PDFDocument {
  constructor(pdfDocProxy) {
    // pdfjs-dist PDFDocumentProxy
    this._pages = [];
    this._metadata = {};
    this._pageCount = 0;
    this.pdfDocProxy = pdfDocProxy;
    this._pageCount = pdfDocProxy.numPages || 0;
  }
  /**
   * Factory method to create PDFDocument from pdfjs-dist PDFDocumentProxy
   */
  static async from(pdfDocProxy) {
    const doc = new _PDFDocument(pdfDocProxy);
    try {
      const metadata = await pdfDocProxy.getMetadata();
      if (metadata) {
        doc._metadata = parseMetadata(metadata);
      }
    } catch (e) {
    }
    for (let i = 1; i <= doc._pageCount; i++) {
      const pageProxy = await pdfDocProxy.getPage(i);
      const page = new PDFPage(doc, pageProxy, i);
      doc._pages.push(page);
    }
    return doc;
  }
  /**
   * Get all pages
   */
  get pages() {
    return this._pages;
  }
  /**
   * Get number of pages
   */
  get pageCount() {
    return this._pageCount;
  }
  /**
   * Get document metadata
   */
  get metadata() {
    return this._metadata;
  }
  /**
   * Get internal pdfjs proxy (for advanced use)
   */
  getProxy() {
    return this.pdfDocProxy;
  }
  /**
   * Close the document and clean up resources
   */
  close() {
    if (this.pdfDocProxy && this.pdfDocProxy.destroy) {
      this.pdfDocProxy.destroy();
    }
  }
};
function parseMetadata(metadata) {
  const result = {};
  const info = metadata.info || {};
  if (info.Title) result.title = info.Title;
  if (info.Author) result.author = info.Author;
  if (info.Subject) result.subject = info.Subject;
  if (info.Keywords) result.keywords = info.Keywords;
  if (info.Creator) result.creator = info.Creator;
  if (info.Producer) result.producer = info.Producer;
  if (info.CreationDate) {
    result.creationDate = parseDate(info.CreationDate);
  }
  if (info.ModDate) {
    result.modificationDate = parseDate(info.ModDate);
  }
  for (const [key, value] of Object.entries(info)) {
    if (!["Title", "Author", "Subject", "Keywords", "Creator", "Producer", "CreationDate", "ModDate"].includes(
      key
    )) {
      result[key] = value;
    }
  }
  return result;
}
function parseDate(dateStr) {
  if (dateStr instanceof Date) {
    return dateStr;
  }
  if (typeof dateStr !== "string") {
    return void 0;
  }
  const cleanStr = dateStr.replace(/^D:/, "");
  const year = parseInt(cleanStr.substring(0, 4), 10);
  const month = parseInt(cleanStr.substring(4, 6), 10) || 1;
  const day = parseInt(cleanStr.substring(6, 8), 10) || 1;
  const hour = parseInt(cleanStr.substring(8, 10), 10) || 0;
  const minute = parseInt(cleanStr.substring(10, 12), 10) || 0;
  const second = parseInt(cleanStr.substring(12, 14), 10) || 0;
  if (isNaN(year)) {
    return void 0;
  }
  return new Date(year, month - 1, day, hour, minute, second);
}

// src/pdfplumber.ts
var PDFPlumber = class _PDFPlumber {
  /**
   * Open a PDF from file path or buffer
   */
  static async open(input) {
    let data;
    if (typeof input === "string") {
      data = await readFile(input);
    } else if (input instanceof Buffer) {
      data = new Uint8Array(input);
    } else {
      data = input;
    }
    return _PDFPlumber.fromBuffer(data);
  }
  /**
   * Load PDF from buffer or Uint8Array
   */
  static async fromBuffer(buffer) {
    const uint8array = buffer instanceof Buffer ? new Uint8Array(buffer) : buffer;
    const loadingTask = getDocument({
      data: uint8array,
      useSystemFonts: true,
      standardFontDataUrl: "https://cdn.jsdelivr.net/npm/pdfjs-dist@4.10.38/standard_fonts/"
    });
    try {
      const pdfDocProxy = await loadingTask.promise;
      return PDFDocument.from(pdfDocProxy);
    } catch (error) {
      loadingTask.destroy();
      throw error;
    }
  }
  /**
   * Load PDF from URL
   */
  static async fromURL(url) {
    const response = await fetch(url);
    const buffer = await response.arrayBuffer();
    return _PDFPlumber.fromBuffer(new Uint8Array(buffer));
  }
};
export {
  CroppedPage,
  PDFDocument,
  PDFPage,
  PDFPlumber
};
